app/services/badge_service.rb

```diff
@@ -0,0 +1,34 @@
+class BadgeService
+  
+  def initialize(user, test_passage)
+    @badge_user = user
+    @test_passage = test_passage
+  end
+
+  def build
+    Badge.all.select { |badge| send("#{badge.rule}?", badge.primary_param) }
+  end
+
+  private
+
+  def category_badge?(category)
+    if @test_passage.test.category == category
```

Comment:
Избегай лишней вложенности, когда это имеет смысл. rubocop такое замечает и предлагает поправить
```suggestion
    return false if @test_passage.test.category != category
```